teps/0139-trusted-artifacts.md

```diff
@@ -0,0 +1,477 @@
+- This proposal does not discuss how to expose artifacts outside of a pipeline, even though it sets foundations that could be used to achieve that
+- This proposal does not discuss how to inject artifact as inputs to a pipeline, even though it sets foundations that could be used to achieve that. For instance, one could use a workspace preprovisioned with artifacts and use artifact type params as inputs for a pipeline
+
+### Use Cases
+
+- Extend the chain of trust across Tasks for provenance produced by Tekton Chains based on the `TaskRuns` and `PipelineRuns` executed by Tekton Pipeline
+
+### Requirements
+
+Support the following combinations:
+
+- One producer, one consumer
+- One producer, N consumers, including with write-one, read-many storage class
+- Many producers, one consumer
+- Many producers, many consumers, including with write-one, read-many storage class
+- Fail validation if workspaces (static/runtime) are not fit for artifacts
+- Fail execution if checksum validation fail, surface error to TaskRun/PipelineRun failure reason
+
+## Proposal
+
+In short the proposal is to:
+
+- extend parameters and result types to a new type `artifact`, an object type with a fixed schema
+- extend the `Pod` logic in the controller to inject hashing and checking steps when required
+- extend variable expansion to include `.data.path` for artifact parameters and results
+
+The same pipeline from demo, rewritten after implementation, is shown by this [demo pipeline](https://gist.github.com/afrittoli/7236be5fca524b752c221d2346497bb7):
+
+<details>
+  <summary>Demo Pipeline:</summary>
+
+```yaml=
+apiVersion: tekton.dev/v1
+kind: PipelineRun
+metadata:
+  generateName: trusted-artifacts-sugar
+spec:
+  pipelineSpec:
+    workspaces:
+      - name: artifactStorage  # In this example this is where we store artifacts
+        artifacts: true # this will result in failed validation if the workspace is bound to a readonly backend like a secret
+    tasks:
+      - name: producer
+        taskSpec:
+          results:
+            - name: aFileArtifact
+              type: artifact # inbuilt object schema (path, hash, type)
+              description: An artifact file
+            - name: aFolderArtifact
+              type: artifact # inbuilt object schema (path, hash, type)
+              description: An artifact folder
+          steps:
+            - name: produce-file
+              image: bash:latest
+              script: |
+                #!/usr/bin/env bash
+                # Produce some content. The result "data.path" will resolve to an empty dir
+                date +%s | tee "$(results.aFileArtifact.data.path)/afile.txt"
+                # The controller appends a step that calculates the hash, copies the data to
```

Comment:
With reusable steps, users will have the ability to back artifacts to other locations as well.
The trusted nature of artifacts is related mainly with the ability to compute and verify the hash. I wonder if we should split it up into two steps:
1. compute hash
2. copy to artifact storage (which can be replaced by a reusable step)

And when consuming an artifact, we:
1. copy from artifact storage to local path (again, which can be replaced by a reusable step)
2. verify the hash. 

As you can see, the hash part needs to be there and likely provided by the controller to ensure that the artifact is trusted, otherwise return an error. However, the copy part is something that users might want differently. May be reusable steps will help there. 

WDYT? 

cc @afrittoli @wlynch @pritidesai @jerop

Reply:
Thanks @chitrangpatel - I like your proposal, I'm working on restructuring the TEP accordingly.
I envision that the logic in those steps will be implemented in `golang`, and we could build a binary that can perform all four functions, depending on the need.